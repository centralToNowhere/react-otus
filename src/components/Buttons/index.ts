export { ButtonGameControl } from "./ButtonGameControl";
export { Button } from "./Button";
