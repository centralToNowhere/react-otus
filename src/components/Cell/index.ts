export { Cell } from "./Cell";
export type { ICell, ICellProps } from "./Cell";
