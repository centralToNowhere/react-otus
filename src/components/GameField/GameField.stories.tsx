import React from "react";
import { Story, Meta } from "@storybook/react";
import { Main as GameField, IGameFieldProps } from "./GameField";

export default {
  title: "Components/GameField",
  component: GameField,
  args: {
    cellSize: 40,
    cellsInRow: 35,
    cellsInCol: 7,
    activeCells: [],
  },
} as Meta;

const Template: Story<IGameFieldProps> = (args) => <GameField {...args} />;

export const FieldSmall = Template.bind({});

FieldSmall.args = {
  cellSize: 20,
  cellsInRow: 20,
  cellsInCol: 5,
  activeCells: [],
};

export const FieldMedium = Template.bind({});

FieldMedium.args = {
  cellSize: 35,
  cellsInRow: 30,
  cellsInCol: 10,
  activeCells: [],
};

export const FieldBig = Template.bind({});

FieldBig.args = {
  cellSize: 40,
  cellsInRow: 42,
  cellsInCol: 10,
  activeCells: [],
};
