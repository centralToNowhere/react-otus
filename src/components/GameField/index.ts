export { GameField } from "./GameField";
export {
  gameFieldSlice,
  setActiveCells,
  setActiveCell,
  setInactiveCell,
  resetCells,
  startGame,
  stopGame,
  defaultGameFieldState,
} from "./slice";
export { selectIndexedCells, getIndexedCells } from "./selectors";
export type { IGameFieldProps } from "./GameField";
export type { IGameFieldState } from "./slice";
