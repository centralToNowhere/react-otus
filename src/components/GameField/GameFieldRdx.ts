import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { ICell } from "@/components/Cell";
import { getInitialCells } from "@/utils/CellGenerator";
import { fieldWidth, fieldHeight } from "@/utils";
import { initialState as FieldControlInitialState } from "@/components/Fields/FieldControlRdx";

export interface IGameFieldState {
  activeCells: ICell[];
}

const defaultState: IGameFieldState = {
  activeCells: [],
};

const initialState: IGameFieldState = {
  activeCells: getInitialCells(
    fieldWidth,
    fieldHeight,
    FieldControlInitialState.cellSize
  ),
};

export const gameFieldSlice = createSlice({
  name: "gameField",
  initialState,
  reducers: {
    setActiveCells: (state, action: PayloadAction<ICell[]>) => {
      state.activeCells = action.payload;
    },
    resetCells: (state) => {
      state = defaultState;
      return state;
    },
  },
});

export const { setActiveCells, resetCells } = gameFieldSlice.actions;
