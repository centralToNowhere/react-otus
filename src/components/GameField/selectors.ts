import { createSelector } from "@reduxjs/toolkit";
import { selectCellsInCol, selectCellsInRow } from "@/components/Fields";
import { RootState } from "@/store/redux/store";
import { ICell, IndexedCells } from "@/Cell/Cell";

export const selectGameField = (state: RootState) => {
  return state.gameField;
};

export const getIndexedCells = (
  activeCells: ICell[],
  cellsInRow: number,
  cellsInCol: number
): IndexedCells => {
  const indexedCells: IndexedCells = [];

  indexedCells.length = cellsInCol * cellsInRow;
  indexedCells.fill(0);

  return activeCells.reduce((indexed: IndexedCells, cell) => {
    if (cell.y < cellsInCol && cell.x < cellsInRow) {
      const i = cellsInRow * cell.y + cell.x;
      indexed[i] = 1;
    }

    return indexed;
  }, indexedCells);
};

export const selectIndexedCells = createSelector(
  selectGameField,
  selectCellsInRow,
  selectCellsInCol,
  (gameField, cellsInRow, cellsInCol): IndexedCells => {
    return getIndexedCells(gameField.activeCells, cellsInRow, cellsInCol);
  }
);

export const selectGenerationNumber = createSelector(
  selectGameField,
  (gameField) => {
    return gameField.generations;
  }
);
