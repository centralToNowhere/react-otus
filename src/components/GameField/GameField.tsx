import React, {
  createRef,
  Dispatch,
  MouseEventHandler,
  MutableRefObject,
} from "react";
import styled from "@emotion/styled";
import { COLORS } from "@/styles/ui-styled";
import { ICell, IndexedCells } from "@/Cell/Cell";
import { AnyAction } from "redux";
import { connect, ConnectedProps } from "react-redux";
import { RootState } from "@/store/redux/store";
import { selectCellsInCol, selectCellsInRow } from "@/components/Fields";
import { selectIndexedCells } from "@/components/GameField/selectors";
import { setActiveCell, setInactiveCell } from "@/components/GameField/slice";
import {
  selectFigures,
  selectPaletteActive,
  selectPaletteCurrent,
} from "@/components/FigurePalette/selectors";

const mapStateToProps = (state: RootState) => {
  return {
    cellSize: state.fieldControl.cellSize,
    indexedCells: selectIndexedCells(state),
    cellsInRow: selectCellsInRow(state),
    cellsInCol: selectCellsInCol(state),
    paletteFigures: selectFigures(state),
    figurePaletteActive: selectPaletteActive(state),
    currentFigureIndex: selectPaletteCurrent(state),
  };
};

const mapDispatchToProps = (dispatch: Dispatch<AnyAction>) => ({
  setActiveCell: (cell: ICell) => {
    dispatch(setActiveCell(cell));
  },
  setInactiveCell: (cell: ICell) => {
    dispatch(setInactiveCell(cell));
  },
});

const connector = connect(mapStateToProps, mapDispatchToProps);
export type IGameFieldProps = ConnectedProps<typeof connector>;

type IGameFieldState = {
  figureAreaCells: number[];
};

export class Main extends React.PureComponent<
  IGameFieldProps,
  IGameFieldState
> {
  constructor(props: IGameFieldProps) {
    super(props);
  }

  state: IGameFieldState = {
    figureAreaCells: [],
  };

  mouseMoveTimeout: ReturnType<typeof setTimeout> | null = null;
  mouseMoveDelay = 50;

  private readonly gameFieldRef: MutableRefObject<HTMLDivElement | null> =
    createRef<HTMLDivElement | null>();

  renderCells = () => {
    const cellsArr: IndexedCells = [];
    const figureAreaCells = this.state.figureAreaCells;
    cellsArr.length = this.props.cellsInCol * this.props.cellsInRow;
    cellsArr.fill(0);

    return Object.assign(cellsArr, this.props.indexedCells).map(
      (cellState, i) => {
        const className = `cell${cellState ? " cell-active" : ""}${
          figureAreaCells.includes(i) ? " cell-highlight" : ""
        }`;

        return (
          <div
            className={className}
            key={i}
            data-testid="cell"
            data-state={!!cellState}
            aria-label={String(i + 1)}
          />
        );
      }
    );
  };

  componentDidUpdate(prevProps: IGameFieldProps) {
    if (this.props.figurePaletteActive === prevProps.figurePaletteActive) {
      return;
    }

    if (this.props.figurePaletteActive) {
      this.figurePlacementHighlight();
    } else {
      this.figurePlacementRemoveHighlight();
    }
  }

  onMouseMove = (e: MouseEvent) => {
    const el = e.target as HTMLDivElement;
    const gameFieldEl = this.gameFieldRef.current;

    if (!gameFieldEl) {
      return;
    }

    const ariaLabel = el.getAttribute("aria-label");

    if (ariaLabel === null) {
      return;
    }

    const index = parseInt(ariaLabel) - 1;

    if (Number.isNaN(index)) {
      return;
    }

    this.setState({
      figureAreaCells: [index],
    });

    if (this.mouseMoveTimeout) {
      clearTimeout(this.mouseMoveTimeout);
    }

    this.mouseMoveTimeout = setTimeout(() => {
      const figureIndex = this.props.currentFigureIndex;
      const figure = this.props.paletteFigures[figureIndex];

      let highlightIndex = index - 1;
      const indexArray = [];

      for (let i = 0; i < figure.cellsInCol; i++) {
        for (let j = 0; j < figure.cellsInRow; j++) {
          highlightIndex += 1;
          indexArray.push(highlightIndex);
        }

        highlightIndex += this.props.cellsInRow - figure.cellsInRow;
      }

      this.setState({
        figureAreaCells: indexArray,
      });

      gameFieldEl.addEventListener("mouseout", this.onMouseOut, false);
    }, this.mouseMoveDelay);
  };

  onMouseOut = () => {
    if (this.mouseMoveTimeout) {
      clearTimeout(this.mouseMoveTimeout);
    }

    this.setState({
      figureAreaCells: [],
    });
  };

  figurePlacementHighlight() {
    const gameFieldEl = this.gameFieldRef.current;

    if (!gameFieldEl) {
      return;
    }

    gameFieldEl.addEventListener("mousemove", this.onMouseMove, false);
    gameFieldEl.addEventListener("mouseout", this.onMouseOut, false);
  }

  figurePlacementRemoveHighlight() {
    const gameFieldEl = this.gameFieldRef.current;

    if (!gameFieldEl) {
      return;
    }

    gameFieldEl.removeEventListener("mousemove", this.onMouseMove, false);
    gameFieldEl.removeEventListener("mouseout", this.onMouseOut, false);

    this.setState({
      figureAreaCells: [],
    });
  }

  onFigurePlacement(topLeft: ICell) {
    const figureIndex = this.props.currentFigureIndex;
    const figure = this.props.paletteFigures[figureIndex];
    const offsetX = topLeft.x;
    const offsetY = topLeft.y;
    const figureCellsInRow = figure.cellsInRow;

    figure.indexedCells.forEach((cellState, i) => {
      const cell = {
        x: (i % figureCellsInRow) + offsetX,
        y: Math.floor(i / figureCellsInRow) + offsetY,
      };

      if (cellState) {
        this.props.setActiveCell(cell);
      } else {
        this.props.setInactiveCell(cell);
      }
    });
  }

  onCellToggle(state: string | undefined, cell: ICell) {
    if (state === "true") {
      this.props.setInactiveCell(cell);
      return;
    }

    this.props.setActiveCell(cell);
  }

  onCellClick: MouseEventHandler<HTMLDivElement> = (e) => {
    const el = e.target as HTMLDivElement;
    const state = el.dataset.state;
    const number = Number(el.getAttribute("aria-label"));

    const i = Math.ceil(number / this.props.cellsInRow) - 1;
    const j = number - i * this.props.cellsInRow - 1;
    const cell: ICell = {
      x: j,
      y: i,
    };

    if (this.props.figurePaletteActive) {
      this.onFigurePlacement(cell);
      return;
    }

    this.onCellToggle(state, cell);
  };

  render() {
    return (
      <StyledGameField
        onClick={this.onCellClick}
        className={"game-field"}
        cellSize={this.props.cellSize}
        cellsInRow={this.props.cellsInRow}
        cellsInCol={this.props.cellsInCol}
        data-testid={"field"}
        ref={this.gameFieldRef}
      >
        {this.renderCells()}
      </StyledGameField>
    );
  }
}

export const GameField = connector(Main);

interface IFieldStyledContainerProps {
  cellSize: number;
  cellsInRow: number;
  cellsInCol: number;
}

const StyledGameField = styled.div<IFieldStyledContainerProps>`
  width: ${(props) => props.cellsInRow * props.cellSize}px;
  cursor: pointer;
  height: ${(props) => props.cellsInCol * props.cellSize}px;
  display: flex;
  flex-wrap: wrap;
  background: ${COLORS.cellBg};
  margin: 0 auto;

  .cell {
    transition: ${(props) =>
      props.cellsInRow * props.cellsInCol < 1000
        ? "0.5s ease background"
        : "none"};
    width: ${(props) => (props.cellSize ? props.cellSize : 10)}px;
    height: ${(props) => (props.cellSize ? props.cellSize : 10)}px;
    line-height: ${(props) => props.cellSize}px;
    text-align: center;
    border: ${(props) =>
      props.cellSize > 2 ? `1px solid ${COLORS.secondary}` : "none"};
  }

  .cell-active {
    background: ${COLORS.activeCellBg};
  }

  .cell-highlight {
    background: ${COLORS.accent};
  }
`;
