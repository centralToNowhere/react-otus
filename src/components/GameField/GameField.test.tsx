import React from "react";
import { render, screen, waitFor } from "@/utils/test-utils";
import { GameField } from "@/components/GameField";
import userEvent from "@testing-library/user-event/dist";
import { COLORS } from "@/styles/ui-styled";
import { getInitialCells } from "@/utils/CellGenerator";

const initialState = {
  fieldControl: {
    cellSize: 40,
    maxFieldWidth: 600,
    maxFieldHeight: 400,
    capacity: 50,
    speed: 2,
  },
  gameField: {
    activeCells: getInitialCells(600, 400, 40),
    gameInProgress: false,
    generations: 0,
  },
};

describe("GameField tests", () => {
  it("should render game field", () => {
    const { asFragment } = render(<GameField />, {
      preloadedState: initialState,
    });

    const field = screen.getByTestId("field");

    expect(field).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render 150 inactive cells", () => {
    render(<GameField />, {
      preloadedState: {
        ...initialState,
        fieldControl: {
          cellSize: 40,
          maxFieldWidth: 600,
          maxFieldHeight: 400,
          capacity: 50,
          speed: 2,
        },
        gameField: {
          ...initialState.gameField,
          activeCells: [],
        },
      },
    });

    const cells: HTMLElement[] = screen.getAllByTestId("cell");
    expect(cells.length).toBe(150);

    cells.forEach((cell) => {
      expect(cell).not.toHaveStyle({
        background: COLORS.activeCellBg,
      });
    });
  });

  describe("gameField css styles", () => {
    describe("gameField animation", () => {
      it("cells amount < 1000 - transition: 0.5s ease background", () => {
        render(<GameField />, {
          preloadedState: {
            ...initialState,
            fieldControl: {
              ...initialState.fieldControl,
              cellSize: 10,
              maxFieldWidth: 100,
              maxFieldHeight: 100,
            },
          },
        });

        const cells = screen.getAllByTestId("cell");

        expect(cells[0]).toHaveStyle({
          transition: "0.5s ease background",
        });
      });

      it("cells amount > 1000 - transition: none", () => {
        render(<GameField />, {
          preloadedState: {
            ...initialState,
            fieldControl: {
              ...initialState.fieldControl,
              cellSize: 10,
              maxFieldWidth: 200,
              maxFieldHeight: 1000,
            },
          },
        });

        const cells = screen.getAllByTestId("cell");

        expect(cells[0]).toHaveStyle({
          transition: "none",
        });
      });

      it("cells amount === 1000 - transition: none", () => {
        render(<GameField />, {
          preloadedState: {
            ...initialState,
            fieldControl: {
              ...initialState.fieldControl,
              cellSize: 10,
              maxFieldWidth: 1000,
              maxFieldHeight: 100,
            },
          },
        });

        const cells = screen.getAllByTestId("cell");

        expect(cells[0]).toHaveStyle({
          transition: "none",
        });
      });
    });

    describe("gameField border", () => {
      it(`cells size > 2 - border: 1px solid ${COLORS.secondary}`, () => {
        render(<GameField />, {
          preloadedState: {
            ...initialState,
            fieldControl: {
              ...initialState.fieldControl,
              cellSize: 3,
              maxFieldWidth: 10,
              maxFieldHeight: 10,
            },
          },
        });

        const cells = screen.getAllByTestId("cell");

        expect(cells[0]).toHaveStyle({
          border: `1px solid ${COLORS.secondary}`,
        });
      });

      it(`cells size === 2 - border: none`, () => {
        render(<GameField />, {
          preloadedState: {
            ...initialState,
            fieldControl: {
              ...initialState.fieldControl,
              cellSize: 2,
              maxFieldWidth: 10,
              maxFieldHeight: 10,
            },
          },
        });

        const cells = screen.getAllByTestId("cell");

        expect(cells[0]).toHaveStyle({
          border: "none",
        });
      });

      it(`cells size < 2 - border: none`, () => {
        render(<GameField />, {
          preloadedState: {
            ...initialState,
            fieldControl: {
              ...initialState.fieldControl,
              cellSize: 1,
              maxFieldWidth: 10,
              maxFieldHeight: 10,
            },
          },
        });

        const cells = screen.getAllByTestId("cell");

        expect(cells[0]).toHaveStyle({
          border: "none",
        });
      });
    });
  });

  describe("cell toggle tests", () => {
    it("should toggle cell inactive -> active", async () => {
      render(<GameField />, {
        preloadedState: {
          ...initialState,
          fieldControl: {
            ...initialState.fieldControl,
            cellSize: 10,
            maxFieldWidth: 100,
            maxFieldHeight: 100,
          },
          gameField: {
            ...initialState.gameField,
            activeCells: [],
          },
        },
      });

      const cells = screen.getAllByTestId("cell");

      userEvent.click(cells[44]);

      await waitFor(() => {
        expect(cells[44]).toHaveStyle({
          background: COLORS.activeCellBg,
        });
      });
    });

    it("should toggle cell active -> inactive", async () => {
      render(<GameField />, {
        preloadedState: {
          ...initialState,
          fieldControl: {
            ...initialState.fieldControl,
            cellSize: 10,
            maxFieldWidth: 100,
            maxFieldHeight: 100,
          },
          gameField: {
            ...initialState.gameField,
            activeCells: [
              {
                x: 4,
                y: 4,
              },
            ],
          },
        },
      });

      const cells = screen.getAllByTestId("cell");

      expect(cells[44]).toHaveStyle({
        background: COLORS.activeCellBg,
      });

      userEvent.click(cells[44]);

      await waitFor(() => {
        expect(cells[44]).not.toHaveStyle({
          background: COLORS.activeCellBg,
        });
      });
    });
  });
});
