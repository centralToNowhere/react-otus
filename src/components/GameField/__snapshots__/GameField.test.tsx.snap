// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameField tests should render game field 1`] = `
<DocumentFragment>
  .emotion-0 {
  width: 600px;
  cursor: pointer;
  height: 400px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  background: rgb(193,184,244);
  margin: 0 auto;
}

.emotion-0 .cell {
  -webkit-transition: 0.5s ease background;
  transition: 0.5s ease background;
  width: 40px;
  height: 40px;
  line-height: 40px;
  text-align: center;
  border: 1px solid #e8d5f0;
}

.emotion-0 .cell-active {
  background: rgb(144, 131, 230);
}

<div
    class="game-field emotion-0 emotion-1"
    data-testid="field"
  >
    <div
      aria-label="1"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="2"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="3"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="4"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="5"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="6"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="7"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="8"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="9"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="10"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="11"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="12"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="13"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="14"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="15"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="16"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="17"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="18"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="19"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="20"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="21"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="22"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="23"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="24"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="25"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="26"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="27"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="28"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="29"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="30"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="31"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="32"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="33"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="34"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="35"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="36"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="37"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="38"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="39"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="40"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="41"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="42"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="43"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="44"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="45"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="46"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="47"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="48"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="49"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="50"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="51"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="52"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="53"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="54"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="55"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="56"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="57"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="58"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="59"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="60"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="61"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="62"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="63"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="64"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="65"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="66"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="67"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="68"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="69"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="70"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="71"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="72"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="73"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="74"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="75"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="76"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="77"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="78"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="79"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="80"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="81"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="82"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="83"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="84"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="85"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="86"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="87"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="88"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="89"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="90"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="91"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="92"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="93"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="94"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="95"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="96"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="97"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="98"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="99"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="100"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="101"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="102"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="103"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="104"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="105"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="106"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="107"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="108"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="109"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="110"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="111"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="112"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="113"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="114"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="115"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="116"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="117"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="118"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="119"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="120"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="121"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="122"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="123"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="124"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="125"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="126"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="127"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="128"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="129"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="130"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="131"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="132"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="133"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="134"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="135"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="136"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="137"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="138"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="139"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="140"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="141"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="142"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="143"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="144"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="145"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="146"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="147"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="148"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
    <div
      aria-label="149"
      class="cell cell-active"
      data-state="true"
      data-testid="cell"
    />
    <div
      aria-label="150"
      class="cell"
      data-state="false"
      data-testid="cell"
    />
  </div>
</DocumentFragment>
`;
